src/socket.cr

```diff
@@ -116,8 +116,11 @@ class Socket < IO
   # sock.bind 1234
   # ```
   def bind(port : Int)
-    Addrinfo.resolve("::", port, @family, @type, @protocol) do |addrinfo|
-      system_bind(addrinfo, "::#{port}") { |errno| errno }
+    domain = @family.inet? ? "0.0.0.0" : "::"
+    domain_and_port = @family.inet? ? "0.0.0.0:#{port}" : "::#{port}"
+
+    Addrinfo.resolve(domain, port, @family, @type, @protocol) do |addrinfo|
+      system_bind(addrinfo, domain_and_port) { |errno| errno }
```

Comment:
I have just this small suggestion to reduce to a single conditional instead of two for optimizing performance - and IMO also legibility.
Also the domain `domain` isn't really appropriate here, it's about addresses, not host names.

```suggestion
    if @family.inet?
      address = "0.0.0.0"
      address_and_port = "0.0.0.0:#{port}"
    else
      address = "::"
      address_and_port = "::#{port}"
    end

    Addrinfo.resolve(address, port, @family, @type, @protocol) do |addrinfo|
      system_bind(addrinfo, address_and_port) { |errno| errno }
```